pubsub/subscriber.go

```diff
@@ -0,0 +1,52 @@
+package pubsub
+
+import (
+	"context"
+
+	"gx/ipfs/QmTu65MVbemtUxJEWgsTtzv9Zv9P8rvmqNA4eG9TrTRGYc/go-libp2p-peer"
+	"gx/ipfs/QmepvmmYNM6q4RaUiwEikQFhgMFHXg2PLhx2E9iaRd3jmS/go-libp2p-pubsub"
+)
+
+// Subscriber subscribes to pubsub topics
+type Subscriber struct {
+	pubsub *pubsub.PubSub
+}
+
+// Message defines the common interface for go-filecoin message consumers.
+// It's a subset of the go-libp2p-pubsub/pubsub.go Message type.
+type Message interface {
+	GetFrom() peer.ID
+	GetData() []byte
+}
+
+// Subscription is a handle to a pubsub subscription.
+// This matches part of the interface to a libp2p.pubsub.Subscription.
+type Subscription interface {
+	// Topic returns this subscription's topic name
+	Topic() string
+	// Next returns the next message from this subscription
+	Next(ctx context.Context) (Message, error)
+	// Cancel cancels this subscription
+	Cancel()
+}
+
+// NewSubscriber builds a new subscriber
+func NewSubscriber(sub *pubsub.PubSub) *Subscriber {
+	return &Subscriber{pubsub: sub}
+}
+
+// Subscribe subscribes to a pubsub topic
+func (s *Subscriber) Subscribe(topic string) (Subscription, error) {
```

Comment:
First note that I've not created a `Subscriber` interface, only a concrete struct. The plumbing holds references to a concrete type, but one owned by us. The `Subscribe` method on that struct does differ from `libp2p`'s, though, omitting a unused variadic parameter of another `libp2p`-defined type (`SubOpt`, an alias for a function pointer). Defining a `Subscriber` interface is possible but not currently needed in order to abstract libp2p from code, excepting `node.go`. It too would need a wrapper to preserve the simpler `Subscribe`, since Go won't accept a function absent an arg as a subtype of a function with a trailing variadic arg.

One level down, `Subscription` is an interface. It matches `libp2p` as closely as possible while hiding the `libp2p/pubsub.Message` type (a struct 😕). The local `Message` interface is a subset of the `libp2p` one, but it seems the type checker doesn't consider a pointer to a struct `Message` acceptable as a value of an interface `Message`, for reasons that make sense to me.

```pubsub/subscriber.go:41:2: cannot use sub (type *pubsub.Subscription) as type Subscription in return argument:
	*pubsub.Subscription does not implement Subscription (wrong type for Next method)
		have Next(context.Context) (*pubsub.Message, error)
		want Next(context.Context) (Message, error)```

However, writing the seemingly-redundant extension struct and wrapper around `Next` gives Go the hook to do an implicit conversion from struct pointer to interface value.

We could expose the libp2p `Message` struct, but that will add libp2p imports in a few places and make use in testing harder (it's basically a wrapper of a protobuf). Or libp2p could define `Message` as an interface, which we could then subset. I think this is an example of the Go pattern of defining interfaces only at consumers falls down; the producer would do us a favour by defining a (broader) interface.